iocore/eventsystem/I_MIOBufferWriter.h

```diff
@@ -0,0 +1,174 @@
+#if !defined I_MIOBUFFERWRITER_H_
+#define I_MIOBUFFERWRITER_H_
+
+/** @file
+
+    Buffer Writer for an MIOBuffer.
+
+    @section license License
+
+    Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+#include <cstring>
+
+#include <ts/ink_assert.h>
+#include <ts/BufferWriter.h>
+
+#if !defined(UNIT_TEST_BUFFER_WRITER)
+#include <I_IOBuffer.h>
+#endif
+
+class MIOBufferWriter : public ts::BufferWriter
+{
+public:
+  MIOBufferWriter(MIOBuffer &miob) : _miob(miob) {}
+
+  MIOBufferWriter &
+  write(const void *data_, size_t length) override
+  {
+    const char *data = static_cast<const char *>(data_);
+
+    while (length) {
+      IOBufferBlock *iobbPtr = _miob.first_write_block();
+
+      if (!iobbPtr) {
+        addBlock();
```

Comment:
Any thoughts on block size allocation strategy? I think it would be better to bump the size up one block size every allocation, to make the overall size growth / allocation more closely match what you'd get with a `std::vector` or similar.

Reply:
I did not realize block size was variable.  But, I think the std::vector approach only makes sense to optimize random access, and MIOBufferWriter only does sequential write access.

Reply:
I think it's a general size optimization strategy for amortizing allocations, independent of the access pattern.

Reply:
I don't see that, do you have a link for that?

Reply:
Suppose the block size is 256 bytes and you write a 3K  string. With the same size there are 12 allocations. With increasing (say double) sizing there are 4 allocations (256, 512, 1024, 2048). The access pattern is irrelevant.

Reply:
No, always add a block that's one size larger than the previously added block. IOBufferBlocks are categorized by a size index, which means each block should be of the next larger size index. In this case, that will mean doubling the block size. See `BUFFER_SIZE_INDEX_...` in `I_IOBuffer.h`.